ring/token_range.go

```diff
@@ -0,0 +1,136 @@
+package ring
+
+import (
+	"math"
+
+	"github.com/pkg/errors"
+	"golang.org/x/exp/slices" // using exp/slices until moving to go 1.21.
+)
+
+// TokenRanges describes token ranges owned by an instance.
+// It consists of [start, end] pairs, where both start and end are inclusive.
+type TokenRanges []uint32
+
+func (tr TokenRanges) IncludesKey(key uint32) bool {
+	switch {
+	case len(tr) == 0:
+		return false
+	case key < tr[0]:
+		// key comes before the first range
+		return false
+	case key > tr[len(tr)-1]:
+		// key comes after the last range
+		return false
+	}
+
+	index, found := slices.BinarySearch(tr, key)
+	switch {
+	case found:
+		// ranges are closed
+		return true
+	case index%2 == 1:
+		// hash would be inserted after the start of a range (even index)
+		return true
+	default:
+		return false
+	}
+}
+
+// GetTokenRangesForInstance returns the token ranges owned by an instance in the ring.
+//
+// Current implementation only works with multizone setup, where number of zones is equal to replication factor.
```

Comment:
I don't know whether it might be of interest for this implementation, but there is a struct called `ringToken` declared in `ownership_priority_queue.go` as:
```
type ringToken struct {
	token     uint32
	prevToken uint32
}
```

Reply:
I wouldn't reuse the type only because it has same two uint32 fields. I was considering to create custom type with similar fields (start/end), but decided to keep implementation as-is for now, as the code is pretty straightforward to follow (imho).